### ISSUE TITLE
How to use with an array of notifications?

### ISSUE THREAD

**Issue by idolize** [pburtchaell/react-notification#30]
The behavior of using `isActive` to *perform the slide in/out transitions* does not work if you render multiple notifications based on an array like so:

```js
return notifications.map(message => 
      <Notification
        isActive={true}
        message={message} />
    );
```
(Instead, the components just mount/unmount instantly and never do the slide in/out animation)

Using `isActive` to show/hide notifications works fine if you only have one or two notifications, like in the example, but I believe in the real world many apps push notifications arbitrarily based on events, rather than keeping some persistent/hidden notifications around at all times.

Is there a way around using `isActive`? (Perhaps with the [`CSSTransitionGroup`](https://facebook.github.io/react/docs/animation.html) addon?)

**Comment by pburtchaell** [comment]
The idea with the component is you would have one component and your message (and action, if it used) values would change for each new notification, e.g., when "notification A" is emitted the component uses the message from "notification A" and when "notification B" is emitted the component uses the message from "notification B".

**Comment by idolize** [comment]
Fair enough; so with this library "stacking" notifications is not really a supported scenario?

**Comment by pburtchaell** [comment]
I am sure you could achieve it with a bit of work/hacking, but I wouldn't say that it is built for that out of the box. I am happy to look at some ideas for how we could change it to work with your question though.

**Comment by meta-meta** [comment]
My solution looks like this. I've since changed the onDismiss in NotificationList to use UIDs as keys instead of index, which allows for duplicate notifications. Sorry about the formatting.

``` jsx

import React, { Component, PropTypes } from 'react';
import Notification from 'react-notification';

class NotificationWrapper extends Component {
    constructor() {
        super();
        this.state = {active: false};
        this.onDismiss = this.onDismiss.bind(this);
    }

    componentDidMount() {
        // react-notification will not animate if mounted with isActive=true
        // mount it with false and wait till next cycle to update it to true
        setTimeout(this.setState.bind(this, {active: true}), 1);
    }

    onDismiss() {
        this.setState({active: false});

        // animation length for react-notification is half a second
        // call this.props.onDismiss so that the parent list can remove this component
        setTimeout(this.props.onDismiss, 500);
    }

    render() {
        return (
                <Notification
                        message={this.props.message}
                        action={this.props.action || 'dismiss'}
                        dismissAfter={this.props.dismissAfter}
                        isActive={this.state.active}
                        onClick={this.props.onClick || this.onDismiss}
                        onDismiss={this.onDismiss}
                        style={{
          active: {
            left: '1em',
          },
          bar: {
            bottom: `${2 + this.props.index * 4}em`,
            padding: '1em',
            font: '1em normal Roboto, sans-serif',
            zIndex: 1,
          },
          action: {
            padding: '0.125em',
            marginLeft: '1em',
            font: '.75em normal Roboto, sans-serif',
            lineHeight: '1em',
          },
        }}
                />
        );
    }
}

NotificationWrapper.propTypes = {
    message: PropTypes.string.isRequired,
    action: PropTypes.string,
    index: PropTypes.number.isRequired,
    onDismiss: PropTypes.func.isRequired,
    onClick: PropTypes.func,
    dismissAfter: PropTypes.number,
};

export default NotificationWrapper;
```

``` jsx
import React, { Component, PropTypes } from 'react';
import NotificationWrapper from 'notificationWrapper';

class NotificationList extends Component {
    constructor() {
        super();
        this.state = {
            msgs: [
                {
    message: 'React is awesome!',
    dismissAfter: 10000,
    key: _.uniqueId(),
  },
  {
    message: 'React is really awesome!',
    dismissAfter: 10000,
    key: _.uniqueId(),
  },
  {
    message: 'jQuery is great!',
    action: 'undo',
    dismissAfter: 10000,
    onClick: () => { alert('React!'); },
    key: _.uniqueId(),
  }
            ],
        };
    }

    render() {
        return (
            <div>
                {
                    this.state.msgs.map((msg, i) => <NotificationWrapper
                        {...msg}
                        index={i}
                        key={`msg${msg.key}`}
                        onDismiss={() => {
                          let nextMsgs = this.state.msgs.slice(); // clone the array
                          nextMsgs.splice(i, 1);
                          this.setState({msgs: nextMsgs});
                        }}
                    />)
                }

            </div>

        );
    }
}

export default NotificationList;
```

**Comment by pburtchaell** [comment]
:+1: I like this approach. 

I would not mind building this as a separate `<NotificationTree />` component that people could optionally use by `import NotificationTree from `react-notification/tree`. Having it as a separate component would be best because we don't want to include all the code for it in the build of people who are just using the single notification component.

If you want to make a PR, please feel free. I'll probably make a few changes after merging it in, but this is good.

**Comment by meta-meta** [comment]
Thank you, I'll have to clear it with my client and refine it a bit. I agree about having it as a separate component.

**Comment by 9mm** [comment]
That would be amazing. This is a cool component but it seems kind of unusable without the traditional stacking of notifications.

**Comment by meta-meta** [comment]
I'll get to this by the end of the week. My client gave the go-ahead.

**Comment by 18601673727** [comment]
:+1:   I'm waiting on this so much.

**Comment by pburtchaell** [comment]
I'm working on it. :smile: I have been a bit slammed with work & final exams at school. I hope to get it released within the next two weeks!

**Comment by pburtchaell** [comment]
This has been released in `react-notification@4.0.0`. 

Please let me know if you have any issues. I eliminated the issues I discovered during the process of preparing this for a release, but I also have not used this in production yet. 

If you need a demonstration of how it works, there is a [working example](https://github.com/pburtchaell/react-notification/blob/master/examples/notification-tree/index.js) that uses the new component combined with Immutable.js.
